pkg/config/config.go

```diff
@@ -19,38 +20,59 @@ type Config struct {
 		Git GitRepositories `glu:"git"`
 		OCI OCIRepositories `glu:"oci"`
 	} `glu:"sources"`
+	Server Server `glu:"server"`
 }
 
-func (c *Config) setDefaults() error {
-	if err := c.Log.setDefaults(); err != nil {
-		return err
-	}
-
-	if err := c.Sources.Git.setDefaults(); err != nil {
-		return err
-	}
+type defaulter interface {
+	setDefaults() error
+}
 
-	if err := c.Sources.OCI.setDefaults(); err != nil {
-		return err
-	}
+func (c *Config) SetDefaults() error {
+	return processStruct(reflect.ValueOf(c).Elem(), func(d defaulter) error {
+		return d.setDefaults()
+	})
+}
 
-	return nil
+type validater interface {
+	validate() error
 }
 
 func (c *Config) Validate() error {
-	if err := c.Log.validate(); err != nil {
-		return err
-	}
+	return processStruct(reflect.ValueOf(c).Elem(), func(v validater) error {
+		return v.validate()
+	})
+}
 
-	if err := c.Sources.Git.validate(); err != nil {
-		return err
-	}
+func processStruct[T any](val reflect.Value, method func(T) error) error {
```

Comment:
Im curious that this works as both `Credentials` and `Sources` are neither struct nor pointer kinds, they should be `reflect.Map` kinds. So I would assume that this fails walks right over the top two fields. Going to run it locally to see if that is the case.

Reply:
Yep, it doesn't run validate or default for the map types.
I will add some suggestions to cover it.

There are couple ways we can solve it.

Reply:
Maybe since we really need to to handle struct. pointers and maps this should be:
```suggestion
func processValue[T any](val reflect.Value, method func(T) error) error {
```